File: javascript/01-arrays-hashing/findVowels.js
/**
 * Problem: Count Vowels
 * Pattern: Arrays & Hashing (string scanning)
 * Time: O(n) | Space: O(1)
 *
 * Count how many vowels (a, e, i, o, u -- case-insensitive) appear in a string.
 *
 * Input:  countVowels('aeiou')
 * Output: 5
 */
function countVowelsType1(string) {
    const vowel= ['a','e','i','o','u'];
    return  string.split('').filter(e => vowel.includes(e)).length;
 }

function countVowels(string) {
   const matchString = string.toLowerCase().match(/[aeiou]/gi);
   return  matchString ? matchString.length : 0;
}

module.exports = countVowels;
module.exports.alternative = countVowelsType1; // approach 2: filter + includes, O(n) same complexity, less efficient in practice (array allocation)

File: python/01-arrays-hashing/find_vowels.py
"""
Problem: Count Vowels
Pattern: Arrays & Hashing (string scanning)
Time: O(n) | Space: O(1)

Count how many vowels (a, e, i, o, u -- case-insensitive) appear in a string.

Input:  count_vowels('aeiou')
Output: 5
"""

def count_vowels_type1(s):
    vowels = set('aeiou')
    return sum(1 for c in s.lower() if c in vowels)


def count_vowels(s):
    return count_vowels_type1(s)  # same approach; Python has no regex-match-count
                                    # shortcut cleaner than the set-based scan


if __name__ == "__main__":
    print(count_vowels('aeiou'))  # 5
